### PR TITLE
refactor: simplify app server creation

### PR DESCRIPTION
**Pull request by clarkdo**
## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [ ] All new and existing tests are passing.

### CODE REVIEW

File: packages/server/src/server.js [hunk 1]
```diff
@@ -183,62 +184,43 @@ export default class Server {
...
-
-      if (socket) {
-        args.path = socket
-      } else {
-        args.port = port || this.options.server.port
-        args.host = host || this.options.server.host
+  normalizeURL(options, server) {
+    const address = server.address()
+    if (options.protocol !== 'socket') {
+      switch (address.address) {
```
Comment (galvez) [inline packages/server/src/server.js]: Here's another pattern I like to use for this:

```
options.host = {
  '127.0.0.1'': 'localhost',
  '0.0.0.0: ip.address()
}[address.address] || options.host
```

**Review by galvez — APPROVED** [review]
LGTM

File: packages/server/src/server.js (hunk 1 above)
Comment (manniL) [reply packages/server/src/server.js]: I usually prefer a lookup object or a map too (instead of switch/case or if cascades)

Comment (clarkdo) [reply packages/server/src/server.js]: Yep, it looks graceful, but `ip.address` will be executed even host is not `0.0.0.0`. 😄

**Comment by pi0** [comment]
@clarkdo @galvez #4259 covers this improvement. I can cherry-pick goodies from here :)

**Comment by pi0** [comment]
- https://github.com/nuxt/nuxt.js/pull/4259/files#diff-9a38d0846ebfd064bbe0e1a373d70c09R189

File: packages/server/src/server.js (hunk 1 above)
Comment (galvez) [reply packages/server/src/server.js]: @clarkdo good point :)

Comment (galvez) [reply packages/server/src/server.js]:
```js
options.host = ({
  '127.0.0.1': () => 'localhost',
  '0.0.0.0': () => ip.address()
}[address.address] || () => options.host)()
```

Suddenly not so graceful anymore 😄

**Comment by clarkdo** [comment]
> @clarkdo @galvez #4259 covers this improvement. I can cherry-pick goodies from here :)

Cool, you can close this one after cherry-pick 😸

**Comment by pi0** [comment]
@clarkdo Sure! Feel free changing anything if needed over there.

**Pull request closed without merging**